pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -225,34 +240,119 @@ func newControllerNode(cfg VSphereConfig) (*VSphere, error) {
+		if cfg.Global.Datacenters != "" {
+			datacenters = cfg.Global.Datacenters
+		} else {
+			// cfg.Global.Datacenter is deprecated, so giving it the last preference.
+			datacenters = cfg.Global.Datacenter
+		}
+
+		vcConfig := VirtualCenterConfig{
+			User:              cfg.Global.User,
+			Password:          cfg.Global.Password,
+			VCenterPort:       cfg.Global.VCenterPort,
+			Datacenters:       datacenters,
+			InsecureFlag:      cfg.Global.InsecureFlag,
+			RoundTripperCount: cfg.Global.RoundTripperCount,
+			WorkingDir:        cfg.Global.WorkingDir,
+		}
+
+		vSphereConn := vclib.VSphereConnection{
+			Username:          vcConfig.User,
+			Password:          vcConfig.Password,
+			Hostname:          cfg.Global.VCenterIP,
+			Insecure:          vcConfig.InsecureFlag,
+			RoundTripperCount: vcConfig.RoundTripperCount,
+			Port:              vcConfig.VCenterPort,
+		}
+		vsphereIns := VSphereInstance{
+			conn: &vSphereConn,
+			cfg:  &vcConfig,
+		}
+		glog.V(4).Infof("vcConfig for VC %s is %+v.", cfg.Global.VCenterIP, vcConfig)
+		glog.V(4).Infof("vSphereConn for VC %s is %+v.", cfg.Global.VCenterIP, vSphereConn)
+
+		vsphereInstanceMap[cfg.Global.VCenterIP] = &vsphereIns
+	} else {
+		for vcServer, vcConfig := range cfg.VirtualCenter {
+			glog.V(4).Infof("Initializing vc server %s and vcConfig %+v", vcServer, vcConfig)
+			if vcServer == "" {
+				glog.Error("vsphere.cong does not have the VirtualCenter IP address specified")
+			}
+			if vcConfig.User == "" {
+				vcConfig.User = cfg.Global.User
+			}
+			if vcConfig.Password == "" {
+				vcConfig.Password = cfg.Global.Password
+			}
+			if vcConfig.VCenterPort == "" {
+				vcConfig.VCenterPort = cfg.Global.VCenterPort
+			}
+			if vcConfig.Datacenters == "" {
+				if cfg.Global.Datacenters != "" {
+					vcConfig.Datacenters = cfg.Global.Datacenters
+				} else {
+					// cfg.Global.Datacenter is deprecated, so giving it the last preference.
+					vcConfig.Datacenters = cfg.Global.Datacenter
+				}
+			}
+			if vcConfig.WorkingDir == "" {
+				vcConfig.WorkingDir = cfg.Global.WorkingDir
+			}
+			if vcConfig.RoundTripperCount == 0 {
+				vcConfig.RoundTripperCount = cfg.Global.RoundTripperCount
+			}
+
+			vSphereConn := vclib.VSphereConnection{
+				Username:          vcConfig.User,
+				Password:          vcConfig.Password,
+				Hostname:          vcServer,
+				Insecure:          vcConfig.InsecureFlag,
+				RoundTripperCount: vcConfig.RoundTripperCount,
+				Port:              vcConfig.VCenterPort,
+			}
+			vsphereIns := VSphereInstance{
+				conn: &vSphereConn,
+				cfg:  vcConfig,
+			}
+			glog.V(4).Infof("vSphereConn for VC %s is %+v.", vcServer, vSphereConn)
+			glog.V(4).Infof("vcConfig for VC %s is %+v.", vcServer, vcConfig)
+
+			vsphereInstanceMap[vcServer] = &vsphereIns
+		}
 	}
 
 	vs := VSphere{
-		conn: &vSphereConn,
-		cfg:  &cfg,
+		vsphereInstanceMap: vsphereInstanceMap,
+		nodeManager: &NodeManager{
+			vsphereInstanceMap: vsphereInstanceMap,
+		},
+		cfg: &cfg,
```

Comment:
The existing VCP code uses VSphere.cfg heavily. Since we are making incremental changes, I did not want to remove this. Also note that VSphereInstance does not contain some of the fields in VSphereConfig like Network and DIsk.